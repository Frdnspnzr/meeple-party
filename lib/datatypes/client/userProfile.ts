import { Game } from "@/lib/datatypes/client/game";

export enum UserRole {
  USER,
  PREMIUM,
  FRIENDS_AND_FAMILY,
  ADMIN,
}

export interface UserProfile {
  id: string;
  name: string;
  role: UserRole;
  realName?: string;
  about?: string;
  place?: string;
  image?: string;
  bggName?: string;
  favorites: Game[];
}

export interface MyUserProfile extends UserProfile {
  email: string;
  profileComplete: boolean;
}
