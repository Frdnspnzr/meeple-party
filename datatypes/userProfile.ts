import { User } from "@prisma/client";

export type UserProfile = Omit<
  User,
  "email" | "emailVerified" | "profileComplete" | "preferences"
>;

export type UserPreferences = {
  sendAnalyticsData: boolean;
  showRealNameInProfile: boolean;
  showPlaceInProfile: boolean;
  allowSearchByPlace: boolean;
  pageLanguage: string;
  gameLanguage: string;
};

export const defaultUserPreferences: UserPreferences = {
  sendAnalyticsData: false,
  showRealNameInProfile: false,
  showPlaceInProfile: false,
  allowSearchByPlace: false,
  pageLanguage: "auto",
  gameLanguage: "follow",
};
