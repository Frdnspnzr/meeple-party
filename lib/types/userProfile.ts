import { GameLanguage, PageLanguage } from "@/i18n/types";
import { User } from "@prisma/client";

export type UserProfile = Omit<
  User,
  "email" | "emailVerified" | "profileComplete" | "preferences"
>;

export type UserPreferences = {
  sendAnalyticsData: boolean;
  showRealNameInProfile: boolean;
  showPlaceInProfile: boolean;
  allowSearchByPlace: boolean;
  hideProfile: boolean;
  pageLanguage: PageLanguage;
  gameLanguage: GameLanguage;
};

export const defaultUserPreferences: UserPreferences = {
  sendAnalyticsData: false,
  showRealNameInProfile: false,
  showPlaceInProfile: false,
  allowSearchByPlace: false,
  hideProfile: false,
  pageLanguage: "auto",
  gameLanguage: "follow",
};
