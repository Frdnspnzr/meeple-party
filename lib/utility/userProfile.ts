import {
  defaultUserPreferences,
  UserPreferences,
} from "@/lib/types/userProfile";
import { Prisma, User } from "@prisma/client";

export function getUserPreferences(user: User): UserPreferences {
  const preferences = user.preferences as Prisma.JsonObject;
  return {
    ...defaultUserPreferences,
    ...preferences,
  };
}
