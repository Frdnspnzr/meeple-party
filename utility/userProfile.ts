import {
  defaultUserPreferences,
  UserPreferences,
} from "@/datatypes/userProfile";
import { Prisma, User } from "@prisma/client";

export function getUserPreferences(user: User): UserPreferences {
  const preferences = user.preferences as Prisma.JsonObject;
  return {
    ...defaultUserPreferences,
    ...preferences,
  };
}
