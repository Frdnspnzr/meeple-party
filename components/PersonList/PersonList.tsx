import Person from "@/components/Person/Person";
import { UserProfile } from "@/datatypes/userProfile";
import { Stack } from "@mantine/core";
import Link from "next/link";
import styles from "./personlist.module.css";

export interface PersonListProps {
  persons: UserProfile[];
}

export default function PersonList({ persons }: Readonly<PersonListProps>) {
  return (
    <Stack>
      {persons.map((f) => (
        <Link href={`app/profile/${f.id}`} key={f.id} className={styles.link}>
          <Person
            name={f.name || ""}
            realName={f.realName || undefined}
            image={f.image || undefined}
          />
        </Link>
      ))}
    </Stack>
  );
}
