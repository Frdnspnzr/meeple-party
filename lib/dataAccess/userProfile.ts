import { MyUserProfile } from "@/lib/datatypes/client/userProfile";
import axios from "axios";

export function getMyUserProfile(): Promise<MyUserProfile> {
  //TODO update to API v2
  return axios
    .get<MyUserProfile>("/api/user")
    .then((response) => response.data);
}
