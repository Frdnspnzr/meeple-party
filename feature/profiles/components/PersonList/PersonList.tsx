import { UserProfile } from "@/lib/types/userProfile";
import { Stack } from "@mantine/core";
import Link from "next/link";
import Person from "../Person/Person";
import styles from "./personlist.module.css";

export interface PersonListProps {
  persons: UserProfile[];
  onClick?: () => void;
}

export default function PersonList({
  persons,
  onClick = () => {},
}: Readonly<PersonListProps>) {
  return (
    <Stack>
      {persons.map((f) => (
        <Link
          href={`app/profile/${f.id}`}
          key={f.id}
          className={styles.link}
          onClick={onClick}
          role="button"
        >
          <Person
            name={f.name || ""}
            realName={f.realName || undefined}
            image={f.image || undefined}
          />
        </Link>
      ))}
    </Stack>
  );
}
