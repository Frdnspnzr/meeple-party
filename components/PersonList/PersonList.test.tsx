import { UserProfile } from "@/datatypes/userProfile";
import {
  generateArray,
  generateArrayWithIndex,
  generateNumber,
  generateUserProfile,
  getUserProfile,
  render,
} from "@/utility/test";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import PersonList from "./PersonList";

jest.mock("@/i18n/client");

describe("Person list", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  beforeEach(() => {
    jest.resetAllMocks();
  });
  it("Matches snapshot", () => {
    const persons: UserProfile[] = generateArrayWithIndex((i) =>
      getUserProfile(i)
    );
    const { container } = render(<PersonList persons={persons} />);
    expect(container).toMatchSnapshot();
  });
  it("Calls callback", async () => {
    const user = userEvent.setup();

    const persons: UserProfile[] = generateArray(generateUserProfile);
    let called = false;
    const callback = () => {
      called = true;
    };
    render(<PersonList persons={persons} onClick={callback} />);
    const buttons = screen.getAllByRole("button");
    await user.click(buttons[generateNumber(0, buttons.length - 1)]);
    expect(called).toBeTruthy();
  });
});
