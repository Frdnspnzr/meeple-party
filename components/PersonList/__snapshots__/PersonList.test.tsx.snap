// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person list Matches snapshot 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_6d731127 mantine-Stack-root"
    style="--stack-gap: var(--mantine-spacing-md); --stack-align: stretch; --stack-justify: flex-start;"
  >
    <a
      class="link"
      href="app/profile/static-profile-0"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-0"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-0"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-0
        </span>
        <span
          class="realName"
        >
          realName-0
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-1"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-1"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-1"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-1
        </span>
        <span
          class="realName"
        >
          realName-1
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-2"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-2"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-2"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-2
        </span>
        <span
          class="realName"
        >
          realName-2
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-3"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-3"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-3"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-3
        </span>
        <span
          class="realName"
        >
          realName-3
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-4"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-4"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-4"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-4
        </span>
        <span
          class="realName"
        >
          realName-4
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-5"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-5"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-5"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-5
        </span>
        <span
          class="realName"
        >
          realName-5
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-6"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-6"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-6"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-6
        </span>
        <span
          class="realName"
        >
          realName-6
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-7"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-7"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-7"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-7
        </span>
        <span
          class="realName"
        >
          realName-7
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-8"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-8"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-8"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-8
        </span>
        <span
          class="realName"
        >
          realName-8
        </span>
      </div>
    </a>
    <a
      class="link"
      href="app/profile/static-profile-9"
      role="button"
    >
      <div
        class="container"
      >
        <div
          class="avatar"
        >
          <img
            alt="name-9"
            class="image"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            referrerpolicy="no-referrer"
            src="image-9"
            style="color: transparent;"
            width="40"
          />
        </div>
        <span
          class="name"
        >
          name-9
        </span>
        <span
          class="realName"
        >
          realName-9
        </span>
      </div>
    </a>
  </div>
</div>
`;
